src/gdcdictionary/validators.py

```diff
@@ -0,0 +1,155 @@
+"""JSON validator module for validating user supplied json documents.
+
+Example:
+    .. code-block:: python
+        import gdcdictionary
+
+        # single
+        instance = {'type': 'case', 'disease_type': 'bad disease'}
+        for violation in gdcdictionary.validate(instance):
+           print(violation.message, violation.keys)
+
+        # multiple
+        for violation in gdcdictionary.validate_instances([instance]):
+            ...
+
+"""
+from __future__ import annotations
+
+import logging
+import re
+from typing import Any, List, NamedTuple, Dict, Optional, Iterable
+
+from gdcdictionary import gdcdictionary
+from jsonschema import Draft4Validator
+
+
+logger = logging.getLogger(__name__)
+invalid_property_regex = re.compile(r"('[a-zA-Z_-]+')+")
+_validators: Dict[str, SchemaValidator] = {}
+
+
+def _parse_keys_from_error_message(error_msg: str) -> List[str]:
+    missing_prop = invalid_property_regex.findall(error_msg)
+    return [m.replace("'", "") for m in missing_prop]
+
+
+class SchemaValidationError(NamedTuple):
+    """A single schema violation representation.
+
+    Properties:
+        schema: the name of the json schema e.g. aliquot
+        message: the error message describing the violation
+        keys: the list of keys whose constraints were violated.
+    """
+
+    schema: str
+    message: str
+    keys: List[str]
+
+    @classmethod
+    def from_values(
+        cls, schema: str, message: str, keys: List[str]
+    ) -> SchemaValidationError:
+        if "Additional properties are not allowed" in message:
+            message = f"Key(s) {keys} not a valid property for type '{schema}'"
+        logger.debug(
+            "json schema violation for '%s'",
+            schema,
+            extra={"keys": keys, "message": message},
+        )
+        return SchemaValidationError(schema, message, keys)
+
+
+class SchemaValidator:
+    """Validator for individual schemas in the dictionary.
+
+    Properties:
+        name: name of specific dictionary schema
+        validator: json validator initialized with the schema
+    """
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self.validator = Draft4Validator(gdcdictionary.schema[name])
+
+    def post_validate(self, violations: List[SchemaValidationError]) -> None:
```

Comment:
Do we have specific plans to implement logic later for 'post-validate' and `pre_validate`?

Reply:
I am not aware of any specific plans. The existing validator does leave room for post validation work [here](https://github.com/NCI-GDC/gdcdatamodel/blob/09f7e13da37387a5bb364c0e21467cafb465cdc0/gdcdatamodel/validators/json_validators.py#L61). The idea itself is useful and would provide one way to do partial validation, where certain validation errors can be filter out

Reply:
Perhaps I'm misunderstanding the reason why both these methods are left unimplemented. I don't see that SchemaValidator is an ABC nor do I see it extended anywhere. I'm sure it's just a misunderstanding on my part but if you could provide more context on that I would appreciate it.

Reply:
You are right, it isn't extended by anything and if it extended, implementing those methods are optional as well so they are not made abstract. They are unimplemented because there are no default pre and post validation work that needs to happen while doing basic json validation.
In terms of what it can be used for - partial json validation can be a good example, in that scenario it will be extended and either the pre or post validation methods implemented to do custom things either before validation happens (where we inject values previously saved) or after validation (where we filter out validation failures that can be ignored). It is potentially useful, but nothing (within the center) is using it right now, but not sure if anyone out there might find it useful.

Reply:
Decided to remove it as it does not affect the current implementation, I like the idea of extending the validator but a different PR might be more appropriate for it